In [2]:
# import stuff
import pandas as p
import numpy as n
import seaborn as s

ModuleNotFoundError: No module named 'seaborn'

## Data Set 2; Stock Exchange Data

This is the data set containing the stock market information for 14 unique exchanges over the span of 12-31-1965 to 06-03-2021.

There are 3 csv files, but for this project, we will only be using indexData.csv. This has all the null values and extra data that indexProcessed.csv does not. 

[csv link](https://www.kaggle.com/mattiuzc/stock-exchange-data)


### Useful Stats:
- Size of set ~> 112,457
- Columns ~> 8

In [3]:
stock = p.read_csv('data/indexData.csv', header = None, skipinitialspace = True,
                encoding = 'latin-1', low_memory = False, skiprows=1,
                names = ['Index', 'Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume']
)
p.set_option('display.max_rows', 100)

In [4]:
stock

Index        Date         Open         High          Low        Close  \
0        NYA  1965-12-31   528.690002   528.690002   528.690002   528.690002   
1        NYA  1966-01-03   527.210022   527.210022   527.210022   527.210022   
2        NYA  1966-01-04   527.840027   527.840027   527.840027   527.840027   
3        NYA  1966-01-05   531.119995   531.119995   531.119995   531.119995   
4        NYA  1966-01-06   532.070007   532.070007   532.070007   532.070007   
...      ...         ...          ...          ...          ...          ...   
112452  N100  2021-05-27  1241.119995  1251.910034  1241.119995  1247.069946   
112453  N100  2021-05-28  1249.469971  1259.209961  1249.030029  1256.599976   
112454  N100  2021-05-31  1256.079956  1258.880005  1248.140015  1248.930054   
112455  N100  2021-06-01  1254.609985  1265.660034  1254.609985  1258.579956   
112456  N100  2021-06-02  1258.489990  1263.709961  1258.239990  1263.619995   

          Adj Close       Volume  
0        528.690002          0.0  
1        527.210022          0.0  
2        527.840027          0.0  
3        531.119995          0.0  
4        532.070007          0.0  
...             ...          ...  
112452  1247.069946  379696400.0  
112453  1256.599976  160773400.0  
112454  1248.930054   91173700.0  
112455  1258.579956  155179900.0  
112456  1263.619995  148465000.0  

[112457 rows x 8 columns]

In [5]:
# To determine null values, and how many there are
stock.isna().sum()

Index           0
Date            0
Open         2204
High         2204
Low          2204
Close        2204
Adj Close    2204
Volume       2204
dtype: int64

A) The types of data being sampled are the values of 14 major stock indices from around the world. The tickers are listed below:

- NYA 
- IXIC
- HSI
- 000001.SS
- GSPTSE
- 399001.SZ
- NSEI
- GDAXI
- KS11
- SSMI
- TWII
- J203.JO
- N225
- N100

B) The features given by the data are the date, the open, close, and adjusted close. The high and low daily values as well as the volume are also given. These are general stats for all stock parket indices. 

C) 
### Qualitative & Nomial:
- Index

### Qualitative & Ordinal:
- Date

### Quantitative & Ordinal:
- Open
- High
- Low
- Close
- Adj Close

### Quantitative & Nomial:
- Volume

All *Quantitative* (except for Volume) values are continuous, while all *Qualitative* (Including Volume) values are discrete.

D) There are 2204 total null values in every quantitative column. This is due to potential dates that stocks did not happen, rather than neglect. Dates such as holidays or a certain day the stock market was closed. The ranges are given below with **NYA** being the lowest with 1 and **N225** being the highest with 626. This may make pivoting difficult. 

E) These feautures are chosen because as stated earlier, they are typical generic stock attributes and are very good for data analysis.

F) Some other features may be how many hours of trading were in each day. Some days are cut short for certain reasons. Other than that, not sure what else directly relates to the indices themselves. Maybe leader of the country at that time?

In [6]:
# G) Doing a pivot on this data

# Figuring out which dates have null values
stockNull = stock.isnull()
stockNull = stockNull.any(axis = 1)
stockNulls = stock[stockNull]
stockNulls

Index        Date  Open  High  Low  Close  Adj Close  Volume
289      NYA  1967-02-23   NaN   NaN  NaN    NaN        NaN     NaN
26639    HSI  1987-01-01   NaN   NaN  NaN    NaN        NaN     NaN
26659    HSI  1987-01-29   NaN   NaN  NaN    NaN        NaN     NaN
26660    HSI  1987-01-30   NaN   NaN  NaN    NaN        NaN     NaN
26706    HSI  1987-04-06   NaN   NaN  NaN    NaN        NaN     NaN
...      ...         ...   ...   ...  ...    ...        ...     ...
108316  N100  2005-03-28   NaN   NaN  NaN    NaN        NaN     NaN
108511  N100  2005-12-26   NaN   NaN  NaN    NaN        NaN     NaN
110051  N100  2012-01-02   NaN   NaN  NaN    NaN        NaN     NaN
110643  N100  2014-05-01   NaN   NaN  NaN    NaN        NaN     NaN
112090  N100  2019-12-25   NaN   NaN  NaN    NaN        NaN     NaN

[2204 rows x 8 columns]

In [7]:
#sort the number of nulls
SNcount = p.DataFrame()
SNcount['Index'] = stockNulls[['Index']]
SNcount['Total Nulls'] = 0
SNcount = SNcount.groupby('Index').count()
SNcount = SNcount.sort_values('Total Nulls', ascending = False)

"""This is important because now we will know how many values will have NaN when we do our pivots."""
SNcount

Total Nulls
Index                 
N225               626
HSI                258
GSPTSE             250
000001.SS          172
399001.SZ          168
GDAXI              168
SSMI               159
KS11               152
TWII               141
J203.JO             41
NSEI                35
N100                33
NYA                  1

In [8]:
#taking all non-null values and running a pivot on them
stock_exist = stock.dropna()

#This confirms they are all unique
print(len(stock_exist))
print(len(stock_exist[['Index', 'Date']].drop_duplicates()))

dates = ['2002-12-13', '2002-12-14', '2002-12-15', '2002-12-16', '2002-12-17']

stock_pivot = p.DataFrame()
for i in range(0,len(dates)):
    stock_pivot = stock_exist[['Index', 'Date', 'Adj Close']][stock_exist['Date'] == dates[i]]

stock_pivot.pivot(index = 'Date', columns = 'Index', values = 'Adj Close')

110253
110253


Index         000001.SS    399001.SZ        GDAXI  GSPTSE          HSI  \
Date                                                                     
2002-12-17  1416.095947  2878.537842  3139.969971  6621.5  9715.740234   

Index              IXIC       KS11        N100         N225          NYA  \
Date                                                                       
2002-12-17  1392.050049  704.48999  552.549988  8510.730469  5094.740234   

Index              SSMI         TWII  
Date                                  
2002-12-17  4902.299805  4545.603027

In [61]:
# PIVOT OF DATA FOR THE MONTH OF JANUARY 2021
date = stock_exist[(stock_exist['Date'] >= '2021-01-01') & (stock_exist['Date'] <= '2021-01-31')]

date.pivot(index = 'Date', columns = 'Index', values = 'Adj Close')


Index         000001.SS    399001.SZ        GDAXI       GSPTSE          HSI  \
Date                                                                          
2021-01-04  3502.958008  14827.46973  13726.74023  17527.80078  27472.81055   
2021-01-05  3528.677002  15147.57031  13651.21973  17682.50000  27649.85938   
2021-01-06  3550.876953  15187.61035  13891.96973  17828.09961  27692.30078   
2021-01-07  3576.205078  15356.40039  13968.24023  18027.59961  27548.51953   
2021-01-08  3570.110107  15319.29004  14049.53027  18042.09961  27878.22070   
2021-01-11  3531.498047  15115.37988  13936.66016  17934.50000  27908.22070   
2021-01-12  3608.340088  15460.03027  13925.05957  17985.80078  28276.75000   
2021-01-13  3598.652100  15365.42969  13939.70996  17934.69922  28235.59961   
2021-01-14  3565.905029  15070.12988  13988.70020  17958.09961  28496.85938   
2021-01-15  3566.377930  15031.70020  13787.73047  17909.00000  28573.85938   
2021-01-18  3596.219971  15269.26953  13848.34961  17944.90039  28862.76953   
2021-01-19  3566.379883  15003.99023  13815.05957  17957.40039  29642.27930   
2021-01-20  3583.092041  15223.36035  13921.37012  18014.90039  29962.47070   
2021-01-21  3621.263916  15520.59961  13906.66992  17916.19922  29927.75977   
2021-01-22  3606.750000  15628.73047  13873.96973  17845.90039  29447.84961   
2021-01-25  3624.238037  15710.19043  13643.95020  17906.00000  30159.00977   
2021-01-26  3569.428955  15352.41992  13870.99023  17779.40039  29391.25977   
2021-01-27  3573.340088  15413.83984  13620.45996  17424.40039  29297.52930   
2021-01-28  3505.176025  14913.20996  13665.92969  17657.19922  28550.76953   
2021-01-29  3483.069092  14821.99023  13432.87012  17337.00000  28283.71094   

Index              IXIC      J203.JO         KS11         N100         N225  \
Date                                                                          
2021-01-04  12698.45020  60479.42188  2944.449951  1112.520020  27258.38086   
2021-01-05  12818.95996  60921.37109  2990.570068  1109.560059  27158.63086   
2021-01-06  12740.79004  61857.42188  2968.209961  1120.229980  27055.93945   
2021-01-07  13067.48047  63042.87891  3031.679932  1122.130005  27490.13086   
2021-01-08  13201.98047  63519.17969  3152.179932  1132.099976  28139.02930   
2021-01-11  13036.42969  63759.92969  3148.449951  1124.290039          NaN   
2021-01-12  13072.42969  63535.10156  3125.949951  1122.560059  28164.33984   
2021-01-13  13128.95020  63474.53125  3148.290039  1123.609985  28456.58984   
2021-01-14  13112.63965  63884.76953  3149.929932  1133.199951  28698.25977   
2021-01-15  12998.50000  63549.75000  3085.899902  1121.800049  28519.17969   
2021-01-18          NaN  63713.87109  3013.929932  1122.540039  28242.21094   
2021-01-19  13197.17969  63603.26172  3092.659912  1122.180054  28633.46094   
2021-01-20  13457.25000  64108.51172  3114.550049  1130.550049  28523.25977   
2021-01-21  13530.91016  64174.87891  3160.840088  1129.319946  28756.85938   
2021-01-22  13543.05957  63987.92188  3140.629883  1123.640015  28631.44922   
2021-01-25  13635.99023  64559.85156  3208.989990  1110.130005  28822.28906   
2021-01-26  13626.05957  63923.62891  3140.310059  1115.660034  28546.17969   
2021-01-27  13270.59961  62784.51953  3122.560059  1098.489990  28635.21094   
2021-01-28  13337.16016  63206.71875  3069.050049  1106.680054  28197.41992   
2021-01-29  13070.69043  62472.10156  2976.209961  1085.380005  27663.39063   

Index              NSEI          NYA         SSMI         TWII  
Date                                                            
2021-01-04  14132.90039  14376.70020          NaN  14902.03027  
2021-01-05  14199.50000  14536.53027  10694.08984  15000.03027  
2021-01-06  14146.25000  14788.83984  10747.08008  14983.12988  
2021-01-07  14137.34961  14928.75000  10778.66992  15214.00000  
2021-01-08  14347.25000  14966.83008  10797.99023  15463.95020  
2021-01-11  14484.75000  14937.95996  10870.41992  15557.29981  
20